File: homework/tai_nguyen/lesson7.2/513_find_bottom_left_tree_value.ts
/**
 * @link https://leetcode.com/problems/find-bottom-left-tree-value/
 */
function findBottomLeftValue(root: TreeNode | null): number {
  if (root === null) return -1;

  let res = root;
  const queue: TreeNode[] = [root];

  while (queue.length) {
    const levelLength = queue.length;
    for (let i = 1; i <= levelLength; i++) {
      const node = queue.shift()!;
      if (i === 1) {
        res = node;
      }
      if (node.left) queue.push(node.left);
      if (node.right) queue.push(node.right);
    }
  }

  return res.val;
}

File: homework/thinhbka/lesson-5-graph/513_Find_Bottom_Left_Tree_Value.py
# https://leetcode.com/problems/find-bottom-left-tree-value/
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

#Time O(n)
#Space O(n)
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findBottomLeftValue(self, root: Optional[TreeNode]) -> int:
        ans = 0
        frontier = [root]
        while frontier:
            nxt= []
            ans = frontier[0].val
            for node in frontier:
                if node.left:nxt.append(node.left)
                if node.right:nxt.append(node.right)
            frontier = nxt
        return ans